src/runtime/GlobalObjectBuiltinArray.cpp

```diff
@@ -76,8 +76,51 @@ Value builtinArrayConstructor(ExecutionState& state, Value thisValue, size_t arg
     return array;
 }
 
+#define CHECK_ARRAY_LENGTH(num, limit)                                                                               \
+    if ((num) >= (limit)) {                                                                                          \
+        ErrorObject::throwBuiltinError(state, ErrorObject::TypeError, errorMessage_GlobalObject_InvalidArrayLength); \
+    }
+
+static Value arraySpeciesCreate(ExecutionState& state, Object* originalArray, const size_t length)
+{
+    ASSERT(originalArray != nullptr);
+    // Assert: length is an integer Number >= 0.
+    ASSERT((int64_t)length >= 0);
+    Value originalArrayConstructor = originalArray->get(state, ObjectPropertyName(state.context()->staticStrings().constructor)).value(state, originalArray);
+
+    // Let C be undefined.
+    Value C;
+    // Let isArray be IsArray(originalArray).
+    // If isArray is true, then
+    if (originalArray->isArrayObject() == true) {
+        // Let C be Get(originalArray, "constructor").
+        C = originalArrayConstructor;
+
+        // TODO 9.4.2.3. 6.c. (after Realm is implemented)
+        // If Type(C) is Object, then
+        if (C.isObject() == true) {
+            // Let C be Get(C, @@species).
+            C = C.asObject()->get(state, ObjectPropertyName(state, state.context()->vmInstance()->globalSymbols().species)).value(state, C);
+        }
+    }
+
+    // If C is null, let C be undefined.
+    // If C is undefined, return ArrayCreate(length).
+    if (C.isUndefinedOrNull() == true) {
+        return new ArrayObject(state, static_cast<double>(length));
+    }
+    // If IsConstructor(C) is false, throw a TypeError exception.
+    if (C.isConstructor() == false) {
+        ErrorObject::throwBuiltinError(state, ErrorObject::TypeError, state.context()->staticStrings().Array.string(), false, String::emptyString, errorMessage_GlobalObject_ThisNotConstructor);
+    }
+    // Return Construct(C, <<length>>).
+    Value argv[1] = { Value(length) };
+    return C.asFunction()->call(state, C, 1, argv);
```

Comment:
For `Construct`, `ByteCodeInterpreter::newOperation()` is used instead.
(Our current implementation for `Construct` is imperfect, so I'm working on revising it)
And would you please add internal TCs for newly added Array builtins?